src/App.tsx

```diff
@@ -1,26 +1,395 @@
 
 export const App: React.FC = () => {
-  if (!USER_ID) {
+  const [input, setInput] = useState('');
+  const [filter, setFilter] = useState(Filter.All);
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [loading, setLoading] = useState(false);
+  const [loadingTodoIds, setLoadingTodoIds] = useState<number[]>([]);
+  const [isSubmitting, setIsSubmitting] = useState(false);
+  const [isEditing, setIsEditing] = useState(false);
+  const [errorMessage, setErrorMessage] = useState('');
+  const [tempTodo, setTempTodo] = useState<Todo | null>(null);
+  const allChecked = todos.every((todo: Todo) => todo.completed);
+  const [isActive, setIsActive] = useState(allChecked);
+  const [serverError, setServerError] = useState(false);
+  const titleField = useRef<HTMLInputElement>(null);
+  const completedTodos = todos.filter((todo: Todo) => todo.completed);
+  const activeTodos = todos.filter((todo: Todo) => !todo.completed);
+  const todosToToggle = todos.filter(todo => todo.completed === isActive);
+
+  useEffect(() => {
+    setLoading(true);
+    todoService
+      .getTodos()
+      .then(todosFromServer => {
+        wait(1000);
+        setTodos(todosFromServer);
+      })
+      .catch(() => {
+        setErrorMessage(Error.UnableToLoadAll);
+      })
+      .finally(() => setLoading(false));
+  }, []);
+
+  useEffect(() => {
+    if (titleField.current) {
+      return titleField.current.focus();
+    }
+  }, [isSubmitting]);
+
+  useEffect(() => {
+    setIsActive(allChecked);
+  }, [allChecked]);
+
+  useEffect(() => {
+    let timeoutId = 0;
+
+    if (errorMessage) {
+      timeoutId = window.setTimeout(() => {
+        setErrorMessage('');
+      }, 3000);
+    }
+
+    return () => clearTimeout(timeoutId);
+  }, [errorMessage]);
+
+  if (!todoService.USER_ID) {
     return <UserWarning />;
   }
 
+  const handleFilterChange = (newFilter: Filter) => {
+    setFilter(newFilter);
+  };
+
+  // eslint-disable-next-line react-hooks/rules-of-hooks
+  const filterTodos = useCallback(
+    (currentTodos: Todo[], currentFilter: Filter) => {
+      if (currentFilter === Filter.Active) {
+        return currentTodos.filter((todo: Todo) => !todo.completed);
+      } else if (currentFilter === Filter.Completed) {
+        return currentTodos.filter((todo: Todo) => todo.completed);
+      } else {
+        return currentTodos;
+      }
+    },
+    [],
+  );
+
+  // eslint-disable-next-line react-hooks/rules-of-hooks
+  const filteredTodos = useMemo(() => {
+    return filterTodos(todos, filter);
+  }, [filterTodos, filter, todos]);
+
+  const handleTitleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
+    setInput(event.target.value);
+    setErrorMessage('');
+  };
+
+  const submit = (event: React.FormEvent) => {
+    event.preventDefault();
+
+    if (!input.trim()) {
+      setErrorMessage(Error.EmptyTitle);
+
+      return;
+    }
+
+    setIsSubmitting(true);
+    setErrorMessage('');
```

Comment:
is it necessary? you do it in handleTitleChange

Reply:
yes, it is required in tests